ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Node/HasPartialProtocolParams.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeFamilies      #-}
+
+-- | Complete protocol parameters from partial protocol parameters.
+module Ouroboros.Consensus.Cardano.Node.HasPartialProtocolParams (
+    HasPartialProtocolParams (..)
+  , PartialProtocolParams (..)
+  ) where
+
+import qualified Cardano.Chain.Genesis as Genesis
+import qualified Cardano.Chain.Update as Update
+import           Data.Kind (Type)
+import           Ouroboros.Consensus.Byron.Ledger.Block (ByronBlock)
+import           Ouroboros.Consensus.Byron.Node (ByronLeaderCredentials,
+                     PBftSignatureThreshold, ProtocolParams (..))
+import           Ouroboros.Consensus.Cardano.Node.ProtocolVersions
+                     (ProtocolVersion)
+import qualified Ouroboros.Consensus.Mempool.Capacity as Mempool
+import           Ouroboros.Consensus.Protocol.Praos (Praos)
+import           Ouroboros.Consensus.Protocol.TPraos (TPraos)
+import           Ouroboros.Consensus.Shelley.Eras (AllegraEra, AlonzoEra,
+                     BabbageEra, ConwayEra, MaryEra, ShelleyEra)
+import           Ouroboros.Consensus.Shelley.Ledger.Block (ShelleyBlock)
+import           Ouroboros.Consensus.Shelley.Node (ProtocolParams (..))
+
+{-------------------------------------------------------------------------------
+  Class
+-------------------------------------------------------------------------------}
+
+class HasPartialProtocolParams blk where
+  data PartialProtocolParams blk :: Type
```

Comment:
What is nice about this is that we don't have to change the era-specific `ProtocolParams*` types at all, but eventually, it seems less ad-hoc/boilerplate-y to remove the protocol versions from `ProtocolParams*` and to instead e.g. have some `WithProtocolVersion` wrapper type (the eg all of the instances below would become unnecessary). Maybe another thing for #319, or even for this PR if it is not too invasive? :smile: